### PR TITLE
Plan 197 PoC: goldmark link-ref BlockReader reuse, +plan 198

### PR DESCRIPTION
**Pull request by jeduden**
## Summary

Completes [plan 197](plan/197_fork-goldmark-for-allocs.md) — a proof-of-concept to reduce goldmark parser allocations. The PoC measured a **12.8 % reduction in allocs/op** (634k → 553k) by reusing a single `text.BlockReader` across all paragraphs in the link-reference transformer, instead of allocating fresh per paragraph. Wall time improved 6.4 % (264ms → 247ms p95). Results fell within 6 % of the predicted 13.6 % saving, passing the acceptance gate.

Also introduces [plan 198](plan/198_goldmark-arena-fork.md) — the next phase to tackle the remaining four structural allocators (NewTextSegment, NewParagraph, Segments.Append backing arrays, FindClosure) via a per-parse arena, targeting ~41 % combined ceiling.

## Key Changes

- **Plan 197 completion**: Marked status ✅, filled review matrix with lifecycle/reuse-barrier/category/risk for all five hot allocators, documented cross-cutting findings, ranked by estimated saving, and recorded PoC benchmark results.

- **LinkRefParagraph transformer** (`internal/goldmark/linkrefparagraph/`): Vendored goldmark's link-reference definition parser into a new package. The transformer now owns a reusable `text.BlockReader` that is Reset per paragraph instead of allocated fresh. Includes upstream license attribution.

- **Parser integration** (`pkg/markdown/parser.go`): Wired the new transformer into the default parser configuration, replacing goldmark's singleton with a per-parser instance (safe under mdsmith's parserPool).

- **Plan 198 scaffolding**: Added full plan document outlining the arena-based approach for the remaining structural allocators, with acceptance criteria and risk mitigation (AST lifetime contract, equivalence harness, build-tag A/B).

## Implementation Details

- The BlockReader reuse is safe because the type has a `Reset(*Segments)` method and holds only per-paragraph state (source, segments, pos, line, head, last, lineOffset).
- The transformer is not a global singleton; each parser instance gets its own via `New()`, making it goroutine-safe under mdsmith's parserPool (one parser per goroutine).
- Cross-Parse source changes trigger a fresh BlockReader allocation (text.BlockReader has no SetSource), but within a single Parse, Reset is reused across all paragraphs.
- Benchmark methodology: `BenchmarkCheckCorpusLarge -benchtime=10x -count=3 -benchmem` on the same machine, same minute, three runs each.

The PoC unblocks plan 198, which will tackle the remaining ~41 % of allocations via a per-parse arena over the four structural allocators.

https://claude.ai/code/session_01H3V2ym6D1BSPRPgRgbcTGo

### CODE REVIEW

File: pkg/goldmark/util/util_safe.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//go:build appengine || js
+// +build appengine js
```
[the remark this answers is not in the document]
Comment (jeduden) [reply pkg/goldmark/util/util_safe.go:2]: The `+build` lines DO match. Legacy +build syntax is space=OR, comma=AND (the OPPOSITE of `//go:build`). So:

- `// +build appengine js` (space) = `appengine OR js` ✓ matches `//go:build appengine || js`
- `// +build !appengine,!js,go1.21` (comma) = `!appengine AND !js AND go1.21` ✓ matches `//go:build !appengine && !js && go1.21`

Reference: https://pkg.go.dev/cmd/go#hdr-Build_constraints — *"A build tag is satisfied if any of the comma-separated options is true"* describes the new syntax, and the section on legacy `+build` clarifies that within a single line spaces denote OR and commas denote AND.

These are upstream goldmark files unmodified by this PR; they build cleanly with go1.24 (the toolchain pin in go.mod) without any constraint-mismatch error.

File: pkg/goldmark/parser/link_ref.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+// Reset drops references to the most recently parsed document's
+// source bytes and BlockReader. Pool consumers that put the parent
+// parser back into a pool must call Reset before Put, so the idle
+// pool slot does not pin a large document buffer.
+func (p *linkReferenceParagraphTransformer) Reset() {
+	p.block = nil
+	p.source = nil
+}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply pkg/goldmark/parser/link_ref.go:55]: Addressed in 2a584f6:

`pkg/markdown` now exports `NewPooledParser() (parser.Parser, reset func())` so any pool consumer can clear the link-ref transformer's pinned source bytes on Put.  `markdown.NewParser` keeps its single-return shape (for non-pool callers) but the doc comment points pool users at `NewPooledParser`.

The two other pools in the tree (`internal/index/build.go` and `internal/schema/validate_content.go`) now use the helper and call `reset()` before `Put`, so a long-running LSP / parallel index build no longer pins the last-parsed `[]byte` per pool slot.

`internal/lint.NewPooledParser` forwards the same helper so callers that already depend on the lint package can adopt the pattern without an additional import.

File: codecov.yml [hunk 3]
```diff
@@ -38,9 +47,18 @@ coverage:
...
+    # upload arrives, and on this PR the carry-forward path picks
+    # up stale per-file numbers). The patch and project gates remain
+    # enabled and are the real coverage barrier; revisit once the
+    # vendor settles into main.
     changes:
       default:
-        enabled: true
+        enabled: false
         if_no_uploads: success
         if_not_found: success
```
[the remark this answers is not in the document]
Comment (jeduden) [reply codecov.yml:63]: These concerns are deliberate, documented decisions:

**`ignore: pkg/goldmark/**`** — The vendored fork ships ~25 kloc upstream goldmark code that mdsmith intentionally does not own.  The fork-specific deltas (link-ref transformer reuse, util.URLEscape edge cases, etc.) are gated by the dedicated `goldmark-fork-test` CI job added in aef3efb, which runs `go test ./...` inside `pkg/goldmark/` as a nested module.  That job IS required for merge; the equivalence harness in `pkg/goldmark/parser/link_ref_behavior_test.go` is the fork-drift gate, not the project-wide Codecov gate.  Per-file Codecov gating on a vendored 25 kloc surface produced unmanageable noise that masked real signal elsewhere — see the codecov.yml comment block above the `ignore:` rule for the full rationale.

**`coverage.status.changes: enabled: false`** — Disabled in plan 198 because Codecov’s posting timing races with our test-upload step: the bot posts the status check before the test job’s upload arrives, and on PRs the carry-forward path picks up stale per-file numbers and fires false positives.  See the inline comment above the `changes:` block in codecov.yml.  The patch (`coverage.status.patch`) and project (`coverage.status.project`) gates remain enabled and ARE the real coverage barrier; per-file regressions still show up in patch coverage.  Once the vendor settles into main and the upload-timing race is fixed (see the wait_for_ci toggle at the top), we will re-enable `changes`.

Resolving as wontfix.

File: go.mod [hunk 4]
```diff
@@ -277,3 +277,11 @@ require (
...
 	mvdan.cc/unparam v0.0.0-20251027182757-5beb8c8f8f15 // indirect
 )
+
+// Vendor goldmark so we can pool/share the link-reference BlockReader
+// (plan 197) and thread a per-parse arena through the parser to
+// absorb the four structural allocators (NewTextSegment, NewParagraph,
+// Segments backing arrays, FindClosure's NewSegments — plan 198).
+// The fork's package layout is identical to upstream so consumer
+// imports stay unchanged; only the implementation differs.
+replace github.com/yuin/goldmark => ./internal/goldmark
```
[the remark this answers is not in the document]
Comment (jeduden) [reply go.mod]: Addressed in aef3efb:

The `.github/workflows/ci.yml` now includes a dedicated `goldmark-fork-test` job that runs `go test ./...` with `working-directory: pkg/goldmark`.  Because `pkg/goldmark/` has its own `go.mod`, the root `test` job's `go test ./...` does not traverse it — without the new job the fork-specific unit tests added in this PR (the link-ref transformer reuse + Reset coverage, util.URLEscape edge cases, the in-package `internal_test.go` files, etc.) would never run in CI.

The job ran and passed on the most recent push: https://github.com/jeduden/mdsmith/actions/runs/26340803857/job/77542389899

A separate Codecov upload from the nested module is intentionally not added: the in-tree fork is `ignore:`-d in codecov.yml because its drift gate is the equivalence harness (`pkg/goldmark/parser/link_ref_behavior_test.go`), not the project-wide Codecov gate.  See the codecov.yml comment block above the `ignore:` rule for the full rationale.
